File: Trees/Binary_Search_Tree/Implementation/Python/BinarySearchTree.py
import TreeNode 

class BinarySearchTree:
    def __init__(self, element):
        self.root = TreeNode(element)
        self.left = None
        self.right = None


    def insert(self, x):
        self.root = insertStepTwo(self, x, self.root)

    def insertStepTwo(self, x, i):
        if i == None:
            i = TreeNode(x)
        elif x < i.element:
            i.left = insertStepTwo(self, x, i.left)
        elif x > i.element:
            i.right = insertStepTwo(self, x, i.right)
        #else:
            # throw Exception here 
        return i

    def remove(self, x):
        self.root = removeStepTwo(self, x, self.root)


    def removeStepTwo(self, x, i):
        if i == None:
            # throw Exception here
            print("Error")

        elif x < i.element:
            i.left = removeStepTwo(self, x, i.left)
        
        elif x > i.element:
            i.right = removeStepTwo(self, x, i.right)
        
        elif i.right == None:
            i = i.left  
        
        elif i.left == None:
            i = i.right
        
        else:
            i.left = LasgestInLeft(self, i , i.left)
                
        #else:
            # throw Exception here 
        return i
        


    def LargestInLeft(self, i, j):
        if j.right == None:
            i.element = j.element
            j = j.left

        else:
            j.right = largestInLeft(i , j.right)
        
        return j


    def getHeight(self):
        return getHeightStepTwo(self.root, 0)


    def getHeightStepTwo(i, height):
        if i == None:
            height -= 1
        
        else:
            heightleft = getHeightStepTwo(i.left, height)
            heightright = getHeightStepTwo(i.right, height)
            height = heightleft if (heightleft > heightright) else  heightright

        return height


    def getLargestElement(self):
        tmp = self.root

        while tmp.right != None:
            tmp = tmp.right
        
        return tmp.element


    def preOrder(self):

        print(self.key)
        
        if self.left:
            self.left.preOrder()
        if self.right:
            self.right

File: Trees/Binary_Search_Tree/Implementation/Ruby/BinarySearchTree.rb
require_relative "TreeNode.rb"

class BinarySearchTree
    attr_accessor  :root;
    
    def initialize
        @root = nil
    end

    
    def insert(x)
        @root = insertStepTwo(x, @root)
    end

    def insertStepTwo(x, i)
        if i == nil
            i = TreeNode.new(x)
        
        elsif x < i.element
            i.left = insertStepTwo(x, i.left)
        
        elsif x > i.element
            i.right = insertStepTwo(x, i.right)
        
        else
            # throw Exception here
        end

        return i
        
    end



    def remove(x)
        @root = removeStepTwo(x, @root)
    end

    def removeStepTwo(x, i)
        if i == nil
            # throw Exception here
        
        elsif x < i.element
            i.left = removeStepTwo(x, i.left)
        
        elsif x > i.element
            i.right = removeStepTwo(x, i.right)
        
        elsif i.right == nil
            i = i.left 
        
        elsif x > i.element
            i = i.right 
        
        else
            i.left = largestInLeft(i, i.left)
        
        end

        return i
        
    end



    def largestInLeft(i, j)
        if(j.right == nil)
            i.element = j.element
            j = j.left
        
        elsif
            j.right = largestInLeft(i, j.right)
        
        end

        return j
        
    end



    def getHeight
        return getHeightStepTwo(@root, 0)
    end



    def getHeightStepTwo(i, height)
        if i == nil
            height -= 1
        else
            heightleft = getHeightStepTwo(i.left, height + 1)
            heightright = getHeightStepTwo(i.right, height + 1)
            height = (heightleft > heightright) ? heightleft : heightright
        end

        return height
    end



    def getLargestElement
        node = @root

        while node.right != nil do
            node = node.right
        end

        return node.element
    end


end

